### ISSUE TITLE
Duplicated agent responses (LLM inference + TTS audio)

### ISSUE THREAD

**Issue by andrewjhogue** [livekit/agents#323]
I've noticed that occasionally the agent will generate two distinct responses (LLM inference and TTS audio) for the same user input.

Interestingly, the second LLM inference isn't generated until after the first TTS audio is completed.

Usually, the second LLM inference + response will be generated using the entire user input, with the first inference being generated using a fraction of it (eg, it doesn't always seem to wait for the user to finish, or handle an interruption cleanly).

Another variant of this I've seen is that sometimes the LLM inference seems to get "caught" - eg, the AI will respond to the previous question I asked instead of the current one, but only once I ask the following question.

Will add logs here as repro's happen locally.

Am using this local setup on Macbook Chrome:

```
    assistant = VoiceAssistant(
        vad=silero.VAD(),
        stt=deepgram.STT(),
        llm=openai.LLM(model="gpt-4o"),
        tts=elevenlabs.TTS(voice=DEFAULT_VOICE),
        chat_ctx=initial_ctx,
        fnc_ctx=fnc_ctx,
        allow_interruptions=True,
        debug=True
    )
```

**Comment by ubeytd** [comment]
Exactly, I encountered this as well. Then the agent responds, "Seems like there is a misunderstanding," because it responds to the previous question.

**Comment by theomonnom** [comment]
Hey, can it be because two agents are in the same room?
There is a known bug where two agents connect to the same room when using the "connect --room my_room" cmd

**Comment by andrewjhogue** [comment]
Hey @theomonnom - thanks for the response. I don't think so, right now my code generates a unique room ID every time, and am also setting max_participants = 2 on room creation. So in theory this shouldn't be possible?

**Comment by davidzhao** [comment]
@andrewjhogue @ubeytd can you repro on 0.7.1?

**Comment by ubeytd** [comment]
> @andrewjhogue @ubeytd can you repro on 0.7.1?

Yes, I have experienced this

_Another variant of this I've seen is that sometimes the LLM inference seems to get "caught" - eg, the AI will respond to the previous question I asked instead of the current one, but only once I ask the following question._

e.g: I speak something to AI. (Sentence 1) It completely ignores it, silence.

Then when I continue to speak (Sentence 2), it is replying to Sentence 2.

Then out of nowhere, it is replying to (Sentence 1), then it gets confused and says "Seems like there is a misunderstanding."

**Comment by ubeytd** [comment]
@davidzhao just tested on `livekit-agents~=0.7.2` issue still persists.

Could it be related to Elevenlabs latency? Can it be that STT is arriving late and the voice assistant is pushing the message to the chat context later when it's not relevant anymore?

**Comment by andrewjhogue** [comment]
@davidzhao I can confirm the issue still exists on `livekit-agents~=0.7.2` as well.

I'm also now seeing the below sequence:

1. User Speaks
2. Agent Starts Synthesizing Response
3. User Interrupts Agent with Second Speech
4. Agent Responds to First Message
5. Agent then sends a second, separate response to the Interruption Speech.

**Comment by willsmanley** [comment]
I'm experiencing this as well, following this thread...

as an aside, I know that Retell is using livekit cloud for AI agent phone calls and web calls (but not this livekit-agents library), yet whatever they're doing seems to avoid this issue. intuitively then, i would expect this issue to be caused by some logic on the livekit-agents since they are using all of the same providers, rather than elevenlabs latency as suggested above. see: https://docs.retellai.com/api-references/create-phone-call

**Comment by davidzhao** [comment]
confirming this is a bug in the current voice assistant code. we are making a few improvements in the upcoming version that should resolve this. a beta build will be cut early next week. I'll share a link here for everyone to try.

**Comment by willsmanley** [comment]
@davidzhao yall are either crazy or amazing for releasing such an amazingly useful library as open source. i will be happy to help test out in any way. i'm preparing to launch an AI math teacher tablet app backed by your infra and open source libs. owe yall a huge debt of gratitude.

**Comment by davidzhao** [comment]
@willsmanley really appreciate your kind words! it made my day!

Excited to see what you build with it. Drop by the community and say hello! :)

**Comment by davidzhao** [comment]
Good news, we have a [0.8 pre-release](https://pypi.org/project/livekit-agents/0.8.0.dev3/) that you guys could test. Note that this includes a few breaking changes. You would want to use [this example](https://github.com/livekit/agents/blob/dev/examples/voice-assistant/minimal_assistant.py) to test the behavior.

Due to the breaking change, you'll also need the latest .dev versions of the openai and deepgram plugins

**Comment by IbraheemOmer** [comment]
@davidzhao I am getting the following error after trying the 0.8 `pre-release.`

```
ERROR: Cannot install -r requirements.txt (line 5), -r requirements.txt (line 8), livekit-agents==0.8.0.dev3 and livekit-agents[codecs,images]==0.7.2 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested livekit-agents==0.8.0.dev3
    livekit-plugins-silero 0.6.0.dev2 depends on livekit-agents~=0.7
    livekit-plugins-deepgram 0.6.0.dev2 depends on livekit-agents~=0.7
    livekit-agents[codecs,images] 0.7.2 depends on livekit-agents 0.7.2 (Installed)

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts 
```

**Comment by davidzhao** [comment]
> @davidzhao I am getting the following error after trying the 0.8 `pre-release.`

please see:

```
you'll also need the latest .dev versions of the openai and deepgram plugins
```

**Comment by IbraheemOmer** [comment]
> you'll also need the latest .dev versions of the openai and deepgram plugins

Yeah, the error I shared is after I use the latest .dev

**Comment by davidzhao** [comment]
looks like you are installing the latest requirement.txt.. which has not yet been updated to the latest plugin references. Please use the following
* https://pypi.org/project/livekit-plugins-silero/0.6.0.dev2/
* https://pypi.org/project/livekit-plugins-openai/0.7.0.dev2/
* https://pypi.org/project/livekit-plugins-deepgram/0.6.0.dev2/

**Comment by IbraheemOmer** [comment]
> looks like you are installing the latest requirement.txt.. which has not yet been updated to the latest plugin references. Please use the following
> 
> * https://pypi.org/project/livekit-plugins-silero/0.6.0.dev2/
> * https://pypi.org/project/livekit-plugins-openai/0.7.0.dev2/
> * https://pypi.org/project/livekit-plugins-deepgram/0.6.0.dev2/

I have created my own requirement.txt which has the latest plugin references you shared.

**Comment by davidzhao** [comment]
Got it, we'll retry to reproduce with that. thanks for the report.

**Comment by willsmanley** [comment]
@IbraheemOmer you will want to use `--no-deps` with `pip install` to override the conflict since it is a beta release. The requirements.txt deps should align for the real GA release.

**Comment by willsmanley** [comment]
@davidzhao i got it working with these dep versions, but i had to provide some default values for the silero plugin, otherwise it would throw:

```
vad=silero.VAD.load(
        min_speech_duration=0.05,
        min_silence_duration=0.1,
        padding_duration=0.1,
        max_buffered_speech=60.0,
        activation_threshold=0.25,
        sample_rate=16000,
        force_cpu=True
    )
```

i will also be testing out the original issues (missing responses / duplicated responses) and will update here.

pip list:
```
livekit                  0.12.0.dev1
livekit-agents           0.8.0.dev8
livekit-api              0.6.0
livekit-plugins-deepgram 0.6.0.dev7
livekit-plugins-openai   0.7.0.dev7
livekit-plugins-silero   0.6.0.dev7
livekit-protocol         0.6.0
```

**Comment by willsmanley** [comment]
ahhh! the performance is INCREDIBLY good... no more issues with missing/duplicated responses. i am officially switching off of very expensive/brittle [retell](https://retellai.com) and onto this SDK. 

have you done any work to measure UserUtteranceEnd-to-AIUtteranceBeg latency? subjectively it seems very close to natural human webcall speeds, but might help with marketing to get some firm percentile numbers here.

**Comment by davidzhao** [comment]
AWESOME to hear Will! Yes, the next sprint we'll be working on metrics to quantify latency/performance in every component of the pipeline

**Comment by Huzaifa-EJ** [comment]
@davidzhao I am using the latest Elevenlabs .dev7 version with the new dependencies I am getting this error

{"message": "11labs reported an error: input_timeout_exceeded", "pid": 66090, "job_id": "AJ_9aepKp8vJJnn", "timestamp": "2024-07-25T11:27:27.811394+00:00"}

**Comment by IbraheemOmer** [comment]
> @davidzhao I am using the latest Elevenlabs .dev7 version with the new dependencies I am getting this error
> 
> `{"message": "11labs reported an error: input_timeout_exceeded", "pid": 66090, "job_id": "AJ_9aepKp8vJJnn", "timestamp": "2024-07-25T11:27:27.811394+00:00"}`

I am getting the same Elevenlabs error even with the new stable releases.

**Comment by theomonnom** [comment]
Hey @Huzaifa-EJ @IbraheemOmer, is it happening when you're using the VoiceAssistant?

**Comment by Huzaifa-EJ** [comment]
@theomonnom Yes, I am experiencing this issue while using the VoiceAssistant. I am utilizing ElevenLabs TTS, which seems to be causing the error. It works fine with OpenAI TTS.

**Comment by egoldschmidt** [comment]
We are also seeing this error with Eleven Labs (livekit-plugins-elevenlabs 0.7.0, livekit-agents 0.8.0). Note we have a custom LLM implementation attached to our VoiceAssistant so I thought that might be a factor but noticing all the activity on this thread it seems like it might be a broader issue.

**Comment by egoldschmidt** [comment]
Quick followup here. I have confirmed that just like @Huzaifa-EJ, if I swap Eleven Labs with OpenAI TTS everything immediately works fine, even with my custom LLM implementation. So I think there's an issue with the new 11L plugin.

The experience is that the first few words are read aloud but the input stream appears to become "wedged" and after a short while you see `11labs reported an error: input_timeout_exceeded` in the console.

**Comment by IbraheemOmer** [comment]
> Quick followup here. I have confirmed that just like @Huzaifa-EJ, if I swap Eleven Labs with OpenAI TTS everything immediately works fine, even with my custom LLM implementation. So I think there's an issue with the new 11L plugin.
> 
> The experience is that the first few words are read aloud but the input stream appears to become "wedged" and after a short while you see `11labs reported an error: input_timeout_exceeded` in the console.

Yeah, that's exactly the same problem I am facing with Eleven Labs plugin.

**Comment by theomonnom** [comment]
Hey guys, going to take a look at this issue today

**Comment by IbraheemOmer** [comment]
> Hey guys, going to take a look at this issue today

Thank you for taking action, waiting for the new [Version Packages](https://github.com/livekit/agents/pull/526).

**Comment by onur-yildirim-infinitusai** [comment]
@theomonnom Today I started seeing the issue with elevenlabs plugin. Using the v0.7.14 and this is how I initialize

```
self.agent = VoicePipelineAgent(
            vad=self.ctx.proc.userdata["vad"],
            stt=stt,
            llm=InfinitusDummyLLM(),
            tts=elevenlabs.TTS(
                model='eleven_flash_v2_5',
                voice= elevenlabs.tts.VoiceSettings(
                    stability=1.0,
                    similarity_boost=0,
                    style=0,
                    use_speaker_boost=False
                )
            ),
            chat_ctx=llm.ChatContext(),
            before_tts_cb=self.before_tts_cb,
            before_llm_cb=self.before_llm_cb,
            interrupt_speech_duration=0.3,
            interrupt_min_words=3,
        )
```

![Image](https://github.com/user-attachments/assets/3717cc88-76fe-49d2-af3c-4d4f02b1d1da)

**Comment by danmastrowtrusstai** [comment]
Also on v0.7.14 and experienced the same issue.

**Comment by davidzhao** [comment]
can you please try 0.8.0? we just released a few improvements to elevenlabs TTS

**Comment by kirylkliushkin** [comment]
I am seeing this issue in 1.2.4. Looks like the llm's response is dropped halfway and is not added to the history, and then a new request is being issued, and the response duplicates previous response. I've tried Gemini 2.5 flash and Opus Sonnet 4 and looks like the issue does not depend on the model.

**Comment by jp-paper** [comment]
We are experiencing a similar issue using Gemini 2.5 Flash/Pro with Cartesia as STT and TTS on the latest Livekit Agent version.

**Comment by longcw** [comment]
@jp-paper this issue was created a year ago with a very different version right now, please create a new issue with logs or examples for reproducing.

**Comment by yorrick** [comment]
@jp-paper have you found a solution to this? I'm also experiencing duplicated responses with gpt4o-mini and Haiku4.5
